frontend/app_flowy/packages/appflowy_editor/lib/src/render/action_menu/action_menu.dart

```diff
@@ -0,0 +1,180 @@
+class ActionMenuArena {
+  final Map<Path, ActionMenuArenaMember> _members = {};
+  final Set<Path> _visible = {};
+
+  ActionMenuArena._singleton();
+  static final instance = ActionMenuArena._singleton();
+
+  void add(ActionMenuState menuState) {
+    final member = ActionMenuArenaMember(
+      state: menuState,
+      listener: () {
+        final len = _visible.length;
+        if (menuState.isHover || menuState.isPinned) {
+          _visible.add(menuState.path);
+        } else {
+          _visible.remove(menuState.path);
+        }
+        if (len != _visible.length) {
+          _notifyAllVisible();
+        }
+      },
+    );
+    menuState.addListener(member.listener);
+    _members[menuState.path] = member;
+  }
+
+  void _notifyAllVisible() {
+    for (var path in _visible) {
+      _members[path]?.state.notify();
+    }
+  }
+
+  void remove(ActionMenuState menuState) {
+    final member = _members.remove(menuState.path);
+    if (member != null) {
+      menuState.removeListener(member.listener);
+      _visible.remove(menuState.path);
+    }
+  }
+
+  bool isVisible(Path path) {
+    var sorted = _visible.toList()
+      ..sort(
+        (a, b) => a <= b ? 1 : -1,
+      );
+    return sorted.isNotEmpty && path == sorted.first;
+  }
+}
+
+/// Used to manage the state of each [ActionMenuOverlay].
+class ActionMenuState extends ChangeNotifier {
+  final Path path;
+
+  ActionMenuState(this.path) {
+    ActionMenuArena.instance.add(this);
+  }
+
+  @override
+  void dispose() {
+    ActionMenuArena.instance.remove(this);
+    super.dispose();
+  }
+
+  bool _isHover = false;
+  bool _isPinned = false;
+
+  bool get isPinned => _isPinned;
+  bool get isHover => _isHover;
+  bool get isVisible => ActionMenuArena.instance.isVisible(path);
+
+  set isPinned(bool value) {
+    if (_isPinned == value) {
+      return;
+    }
+    _isPinned = value;
+    notifyListeners();
+  }
+
+  set isHover(bool value) {
+    if (_isHover == value) {
+      return;
+    }
+    _isHover = value;
+    notifyListeners();
+  }
+
+  void notify() {
+    notifyListeners();
+  }
+}
+
+/// The default widget to render an action menu
+class ActionMenuWidget extends StatelessWidget {
+  final List<ActionMenuItem> items;
+
+  const ActionMenuWidget({super.key, required this.items});
+
+  @override
+  Widget build(BuildContext context) {
+    final editorStyle = Theme.of(context).extension<EditorStyle>();
+
+    return Card(
+      color: editorStyle?.selectionMenuBackgroundColor,
+      elevation: 3.0,
+      child: Row(
+        mainAxisSize: MainAxisSize.min,
+        children: items.map((item) {
+          return ActionMenuItemWidget(
+            item: item,
+          );
+        }).toList(),
+      ),
+    );
+  }
+}
+
+class ActionMenuOverlay extends StatelessWidget {
+  final Widget child;
+  final List<ActionMenuItem> items;
+  final Positioned Function(BuildContext context, List<ActionMenuItem> items)?
```

Comment:
How about move this customize function  to `ActionProvider`?

Reply:
Before
```dart
child: ActionMenuOverlay(
          items: builder.actions(context),
          customActionMenuBuilder: customActionMenuBuilder,
          child: builder.build(context),
        ),
```

After
```dart
child: ActionMenuOverlay(
          items: builder.actions(context),
          customActionMenuBuilder: builder.customActionMenuBuilder,
          child: builder.build(context),
        ),

Reply:
The intention behind `customActionMenuBuilder` is that projects using appflowy_editor can easily change all action menu appearances.

If we add `ActionProvider.customActionMenuBuilder`, then again there could be different menu designs.
That's why I believe we shouldn't add this function to `ActionProvider`. If a node widget really needs a custom menu, it can always create it itself.